### PR TITLE
Filters

### PR DESCRIPTION
**Pull request by vladfreel**
FILTERS

### CODE REVIEW

File: app/controllers/employees_controller.rb [hunk 1]
```diff
@@ -3,7 +3,15 @@ class EmployeesController < ApplicationController
   before_action :employees, only: %i[index show]
   before_action :skills, only: %i[index show]
 
-  def index; end
+  # TODO
+  def index
+    @skills = Skill.all
+    @employees = Employee.all
+    @employees = Employee.filter(params.reject{|_, v| v.blank?}.slice(:role, :office, :department))
```
Comment (Tivgres) [inline app/controllers/employees_controller.rb]: Maybe better to use if-else?

File: app/models/concerns/filterable.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module Filterable
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def filter(filtering_params)
+      results = self.where(nil)
+      filtering_params.each do |key, value|
+        results = results.public_send(key, value) if key.present?
+      end
+      results
```
Comment (Tivgres) [inline app/models/concerns/filterable.rb]: May produce nil, it's ok?

File: app/models/employee.rb [hunk 3]
```diff
@@ -14,7 +17,21 @@ class Employee < ApplicationRecord
...
+
+  enum department: { other_department: 0, ruby: 1, php: 2, js: 3 }
+
+  enum office: { managers: 0, ruby_office: 1, central: 2, firsts: 3,
+                 kruglyash: 4, gamers:5, admins: 6, remote: 7, lviv: 8 }
+
+  def self.filter_skills(employee, skills_array)
+    emp = []
+    employee.each do |empl|
+      if empl.skills.collect(&:name).any? {|skill| skills_array.include?(skill) }
```
Comment (Tivgres) [inline app/models/employee.rb]: When u used space after and before { } - use it at any side, not just at end or start.

File: app/models/employee.rb [hunk 4]
```diff
@@ -28,6 +45,7 @@ def image
     super || 'user.png'
   end
 
+
```
Comment (Tivgres) [inline app/models/employee.rb]: Unused line?
I think it's bad for style. Just delete it.

File: spec/factories/employee.rb [hunk 5]
```diff
@@ -1,9 +1,9 @@
 FactoryBot.define do
   factory :employee do
-    first_name Faker::Name.first_name
-    last_name Faker::Name.last_name
+    sequence(:first_name) { |n| "#{n}first_name" }
```
Comment (Tivgres) [inline spec/factories/employee.rb]: `first_name { Faker::Name.first_name }`

File: spec/factories/employee.rb [hunk 6]
```diff
@@ -1,9 +1,9 @@
...
+    sequence(:last_name) { |n| "#{n}last_name" }
```
Comment (Tivgres) [inline spec/factories/employee.rb]: `last_name { Faker::Name.last_name }`

File: spec/factories/employee.rb [hunk 7]
```diff
@@ -1,9 +1,9 @@
...
     password Faker::Internet.password
-    email { Faker::Internet.email }
+    sequence(:email) { |n| "#{n}email@mail.ru" }
```
Comment (Tivgres) [inline spec/factories/employee.rb]: `email { Faker::Internet.email }`

File: app/models/employee.rb [hunk 8]
```diff
@@ -1,7 +1,10 @@
 class Employee < ApplicationRecord
+  include Filterable
   extend FriendlyId
   friendly_id :friendly_name, use: :slugged
-
+  scope :role, -> (role) { where role: role }
+  scope :office, -> (office) { where office: office }
+  scope :department, -> (department) { where department: department }
```
Comment (drkmen) [inline app/models/employee.rb]: please use same code style
where are already defined two scopes, follow their style

File: app/models/employee.rb [hunk 9]
```diff
@@ -14,7 +17,21 @@ class Employee < ApplicationRecord
...
   enum role: { other: 0, programmer: 1, manager: 2, team_lead: 3, admin: 4 }
-  enum department: { ruby: 0, php: 1, js: 2, other_department: 3 }
+
+  enum department: { other_department: 0, ruby: 1, php: 2, js: 3 }
+
+  enum office: { managers: 0, ruby_office: 1, central: 2, firsts: 3,
+                 kruglyash: 4, gamers:5, admins: 6, remote: 7, lviv: 8 }
+
+  def self.filter_skills(employee, skills_array)
+    emp = []
```
Comment (drkmen) [inline app/models/employee.rb]: use `.map` if you need new array

File: app/models/employee.rb [hunk 10]
```diff
@@ -14,7 +17,21 @@ class Employee < ApplicationRecord
...
+    employee.each do |empl|
+      if empl.skills.collect(&:name).any? { |skill| skills_array.include?(skill) }
+        emp.push empl
```
Comment (drkmen) [inline app/models/employee.rb]: don't use one line `if` with out `else` statement

File: app/views/employees/_skills.html.haml [hunk 11]
```diff
@@ -1,5 +1,19 @@
 .section.filters
   %h3 Filters
+  = form_tag '', :method => :get do
+    = select_tag :office, options_for_select(Employee.offices.map {|k, v| [k.humanize.capitalize, v]}), prompt: 'Select office'
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: don't need to call `.capitalize` after `.humanize`

File: app/views/employees/_skills.html.haml [hunk 12]
```diff
@@ -1,5 +1,19 @@
...
+    %br
+    %br
+    = select_tag :role, options_for_select(Employee.roles.map {|k, v| [k.humanize.capitalize, v]}), prompt: 'Select role'
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: same

File: app/views/employees/_skills.html.haml [hunk 13]
```diff
@@ -1,5 +1,19 @@
...
+    %br
+    %br
+    = select_tag :department, options_for_select(Employee.departments.map {|k, v| [k.humanize.capitalize, v]}), prompt: 'Select department'
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: same

File: app/views/employees/_skills.html.haml [hunk 14]
```diff
@@ -1,5 +1,19 @@
 .section.filters
   %h3 Filters
+  = form_tag '', :method => :get do
+    = select_tag :office, options_for_select(Employee.offices.map {|k, v| [k.humanize.capitalize, v]}), prompt: 'Select office'
+    %br
+    %br
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: use `%p` instead if two `%br`

File: spec/controllers/employees_controller_spec.rb [hunk 15]
```diff
@@ -3,6 +3,13 @@
 require 'rails_helper'
 
 RSpec.describe EmployeesController, type: :controller do
+  let(:employee_1) { FactoryBot.create(:employee, office: 1, role: 1, department: 1) }
+  let(:employee_2) { FactoryBot.create(:employee, office: 2, role: 2, department: 1) }
+  let(:employee_3) { FactoryBot.create(:employee, office: 2, role: 3, department: 0) }
+  let(:employee_4) { FactoryBot.create(:employee, office: 3, role: 4, department: 3) }
```
Comment (drkmen) [inline spec/controllers/employees_controller_spec.rb]: DRY

File: spec/controllers/employees_controller_spec.rb [hunk 16]
```diff
@@ -16,6 +23,54 @@
     end
   end
 
+  describe 'GET #index with filters' do
+    before do
+      ResourceSkill.create(skill_id: skill_1.id, employee_id: employee_1.id)
+      ResourceSkill.create(skill_id: skill_1.id, employee_id: employee_2.id)
+      ResourceSkill.create(skill_id: skill_3.id, employee_id: employee_2.id)
+      ResourceSkill.create(skill_id: skill_1.id, employee_id: employee_3.id)
+      ResourceSkill.create(skill_id: skill_2.id, employee_id: employee_4.id)
```
Comment (drkmen) [inline spec/controllers/employees_controller_spec.rb]: DRY

File: app/models/concerns/filterable.rb (hunk 2 above)
Comment (vladfreel) [reply app/models/concerns/filterable.rb]: it's ok

File: app/models/employee.rb [hunk 17]
```diff
@@ -15,6 +18,16 @@ class Employee < ApplicationRecord
 
   enum role: { other: 0, programmer: 1, manager: 2, team_lead: 3, admin: 4 }
   enum department: { ruby: 0, php: 1, js: 2, sys_admins: 3, managers: 4, other_department: 5 }
+  enum office: { managers: 0, ruby_office: 1, central: 2, firsts: 3,
+                 kruglyash: 4, gamers: 5, admins: 6, remote: 7, lviv: 8 }
+
+  def self.filter_skills(employee, skills_array)
+    emp = []
```
Comment (drkmen) [inline app/models/employee.rb]: again, use `.map`

File: app/views/employees/_skills.html.haml [hunk 18]
```diff
@@ -1,8 +1,26 @@
 .section.filters
+<<<<<<< 3ff790fee3e7bc66c52054ccfc943dfac40e2c59
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: ???

File: app/views/employees/_skills.html.haml [hunk 19]
```diff
@@ -1,8 +1,26 @@
...
   %h3.text-center Filters
   - @skills.each do |skill_type, skills|
     %h4= skill_type.humanize
     %p
       - skills.each do |skill|
         = link_to skill.name, '#', class: ''
-    %hr
\ No newline at end of file
+    %hr
+=======
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: ???

File: app/views/employees/_skills.html.haml [hunk 20]
```diff
@@ -1,8 +1,26 @@
...
+    %p
+      = select_tag :department, options_for_select(Employee.departments.map {|k, v| [k.humanize, v]}), prompt: 'Select department'
+    %p
+      - @skills.each do |skill|
+        = skill.name
+        = check_box_tag "skills[]", skill.name
+    %p
+      = submit_tag 'Filter'
+
+>>>>>>> Add filters on role and office
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: ???

File: app/models/employee.rb (hunk 17 above)
Comment (vladfreel) [reply app/models/employee.rb]: Can you give example, how it must be view

File: app/controllers/employees_controller.rb [hunk 21]
```diff
@@ -3,7 +3,14 @@ class EmployeesController < ApplicationController
   before_action :employees, only: %i[index show]
   before_action :skills, only: %i[index show]
 
-  def index; end
+  # TODO
+  def index
+    @skills = Skill.all
+    @employees = Employee.filter(params.reject { |_, v| v.blank? }.slice(:role, :office, :department))
+    if params[:skills]
+      @employees = Employee.filter_skills(@employees, params[:skills])
```
Comment (polyakovigor) [inline app/controllers/employees_controller.rb]: @employees = Employee.filter_skills(@employees, params[:skills]) if params[:skills]

File: app/models/employee.rb [hunk 22]
```diff
@@ -17,16 +17,12 @@ class Employee < ApplicationRecord
   has_many :projects
 
   enum role: { other: 0, programmer: 1, manager: 2, team_lead: 3, admin: 4 }
-  enum department: { ruby: 0, php: 1, js: 2, sys_admins: 3, managers: 4, other_department: 5 }
+  enum department: { ruby: 0, php: 1, js: 2, sys_admins: 3, managerss: 4, other_department: 5 }
```
Comment (drkmen) [inline app/models/employee.rb]: typo

Comment (vladfreel) [reply app/models/employee.rb]: kak est, lyche nichego ne pridumal

File: app/views/employees/_skills.html.haml [hunk 23]
```diff
@@ -1,8 +1,16 @@
 .section.filters
-  %h3.text-center Filters
-  - @skills.each do |skill_type, skills|
-    %h4= skill_type.humanize
+  %h3 Filters
+  = form_tag '', :method => :get do
     %p
-      - skills.each do |skill|
-        = link_to skill.name, '#', class: ''
-    %hr
```
Comment (drkmen) [inline app/views/employees/_skills.html.haml]: this is not a way how to resolve conflicts

File: app/controllers/employees_controller.rb [hunk 24]
```diff
@@ -3,7 +3,14 @@ class EmployeesController < ApplicationController
   before_action :employees, only: %i[index show]
   before_action :skills, only: %i[index show]
 
-  def index; end
+  # TODO
+  def index
+    if params.present?
```
Comment (drkmen) [inline app/controllers/employees_controller.rb]: remove `if` statement
```ruby
return unless params.present?
```

**Pull request merged**